Add routes to lookup resources by id

zephraph opened the pull request:
This will address #485 

The intent is to provide a mechanism for fetching resources by id. I'd failed at previous attempts at this but #798 has made it much more feasible. In [previous conversations](https://drive.google.com/file/d/1t_lQyPycGzExShqFVfL4aEyxMU8gUyii/view) we'd discussed using `/by_id/<resource>/<id>` as the general shape of the route which is represented here. Big shout out to @ahl for #1329 which helped to clean up the organization.

Note that while we've discussed adding PUT and DELETE endpoints as well, I'm not tackling that in this PR.

## Requested feedback

- ~`/by_id/` or `/by-id/`~? I've decided to use the `dash`. We have inconsistent usage of dashes and underscores but there's more of the former than the latter. I've got a [openapi-lint](https://github.com/oxidecomputer/openapi-lint/issues/9) issue open to help us enforce consistency here. 
- Should the resource names in the route stay plural or be changed to be singular?
- I've spread the id implementations to be co-located near the associated GET method. Does that feel right or should I move back to having all the `/by-id/` routes grouped together?

davepacheco commented:
Nice!  In terms of prioritizing review, what's the level of urgency on reviewing this?  Is this blocking work for Tuesday?

zephraph commented on nexus/src/db/lookup.rs:
```diff
@@ -231,6 +231,23 @@ impl<'a> LookupPath<'a> {
+        Image { key: ImageKey::PrimaryKey(Root { lookup_root: self }, id) }
+    }
+
+    /// Select a resource of type Snapshot, identified by its id
+    pub fn snapshot_id(self, id: Uuid) -> Snapshot<'a> {
+        Snapshot { key: SnapshotKey::PrimaryKey(Root { lookup_root: self }, id) }
+    }
```
Snapshots aren't technically real at this point. I can remove all snapshot related changes if others think it would be prudent to do so.

jmpesp replied:
Keep them in, they'll be real enough soon :)

zephraph commented:
I consider this low priority and isn't blocking anything that's critical to have on Tuesday.

zephraph commented on nexus/src/external_api/http_entrypoints.rs:
```diff
@@ -3380,14 +3392,169 @@ async fn sshkeys_delete_key(
+}
+
+
+/// Fetch an organization by id
+#[endpoint {
+    method = GET,
+    path = "/by_id/organizations/{id}",
+    tags = ["organizations"],
+}]
+async fn by_id_organization_get(
```
A few things notes
- I preserved the tags for the resources it maps to. I'm assuming that's what we'd want.
- I left the resource plural in the route to match with how it exists in other parts of the API. It's somewhat weird in that it's only used to give you a single item, but it seemed prudent given this could be used for batching in the future
- I grouped all these together because they share the same parent route despite them returning different resources

zephraph commented on nexus/src/external_api/http_entrypoints.rs:
```diff
@@ -3380,14 +3392,169 @@ async fn sshkeys_delete_key(
+/// Fetch a network interface by id
+#[endpoint {
+    method = GET,
+    path = "/by_id/network-
```
hahaha. You can see where my computer ran out of memory and stopped writing to the file. 🤦‍♂️

karencfv commented:
This is real nice!!! Just popping in to say that looking up an item by ID is crucial imo. For example, when implementing the resource for VPCs in the Terraform provider I was already having to disable name updates https://github.com/oxidecomputer/terraform-provider-oxide-demo/blob/main/oxide/resource_vpc.go#L147-L154

zephraph commented:
From @ahl in chat:

> I was thinking about how this might translate to the CLI. I imagine we'd want to do something like oxide disk get --uuid <uuid> or something as an alternative to specifying the org/project/disk
@zephraph if we wanted to support --uuid for other operations (modify, delete) would the cli just do one call to get the disk by name and then another call to modify or delete it?
it seems like this has the same raciness that @davepacheco alluded to, but perhaps it's no big deal if the initial lookup by uuid succeeds and the subsequent operation fails (i.e. those operations always race with delete anyways...)

I agree that it'd be nice to be able to supply something like `--id` to the regular commands and have that work in place of project/org. It seems like adding `PUT` and `DELETE` support for these endpoints would likely make sense. Aside from API bloat, is there any reason _not_ to add them?

karencfv commented:
I am 100% on board with adding PUT and DELETE. I'd even go as far as saying they are crucial as well. Everything should be identifiable by an immutable identifier imo

ahl commented:
A given entity can be named either by its UUID or by some Name tuple (agreed?). This may be a radical change, but another approach would be to dispense with the paths for the name tuple and instead prefer query/body params. In other words, for a single endpoint we could accept either the UUId or the appropriate collection of names. A GET for a disk, e.g., could take either a UUID or a org/instance/disk Name. This is as opposed to the alternative where every CRUD operation would effective have two versions: one by UUID and one by Name-tuple.

I don't claim to be an HTTP API aficionado so it may be that what I'm suggesting would be ugly or horrible.

From the perspective of the CLI, with either model of the API I imagine we'd give users the option of specifying, say, a disk either way. Approximately:

```
oxide disk -u <uuid> <cmd>
oxide disk -o <org> -p <project> -d <disk> <cmd>

where <cmd> may be: view, update, delete
```

david-crespo commented:
Awesome. I'm with Adam about putting the resource name first.

david-crespo commented on nexus/tests/integration_tests/unauthorized.rs:
```diff
@@ -55,22 +56,29 @@ async fn test_unauthorized(cptestctx: &ControlPlaneTestContext) {
     print!("{}", VERIFY_HEADER);
     for endpoint in &*VERIFY_ENDPOINTS {
-        verify_endpoint(&log, client, endpoint).await;
+        verify_endpoint(&log, client, endpoint, &setup_results).await;
```
I'd rather do the lookup of the result here and pass that in vs. passing in the entire list

zephraph commented:
The authorization tests are still failing. It seems to be happening on `/organizations` which is before it gets to the first `/by-id/` request. It could also be a panic from `/by-id/organizations/{id}`. 

I'm seeing these two messages.

```
In this case, an unauthenticated request to "GET /organizations" returned
    401.  All requests to "PUT /organizations" returned 405.
```

```
thread 'integration_tests::unauthorized::test_unauthorized' panicked at 'called `Result::unwrap()` on an `Err` value: expected status code 403 Forbidden, found 404 Not Found
```

zephraph commented:
Okay, I _finally_ got things worked out. I was having a few problems...

1. In `VerifyEndpoint` I had the `by-id` routes listed as `Visibility::Public` (meaning they yielded a 403 instead of a 404). Switching them to `Visibility::Private` fixed the above error.
2. The creation of project images and snapshots is unimplemented so I changed the `by-id` routes of both to also return unimplemented
3. NICs were tricky b/c technically the NIC is already created. I could've done extra work to create a new NIC, but it didn't feel worth the effort. Instead I modified the `SetupReq` to do either GET or POST and be able to associate that to any `id_routes`. It's likely this will be handy in the future for any resources that aren't directly created.

zephraph commented on nexus/tests/integration_tests/endpoints.rs:
```diff
@@ -872,6 +932,14 @@ lazy_static! {
+            url: "/by-id/images/{id}",
+            visibility: Visibility::Protected,
+            allowed_methods: vec![
+                AllowedMethod::GetUnimplemented,
```
This lookup isn't implemented due to the underlying create/fetch not being implemented. Project images are still stubs at this point.

zephraph commented on nexus/tests/integration_tests/endpoints.rs:
```diff
@@ -893,6 +961,15 @@ lazy_static! {
+            url: "/by-id/snapshots/{id}",
+            visibility: Visibility::Protected,
+            allowed_methods: vec![
+                AllowedMethod::GetUnimplemented,
```
Same as above

zephraph commented on nexus/tests/integration_tests/endpoints.rs:
```diff
@@ -579,6 +579,15 @@ lazy_static! {
         },
+
+        VerifyEndpoint {
+            url: "/by-id/organizations/{id}",
```
Note the `{id}` is important here. That's the only way the id will get substituted.

zephraph commented on nexus/tests/integration_tests/unauthorized.rs:
```diff
@@ -152,69 +182,87 @@ lazy_static! {
             body: serde_json::to_value(&*DEMO_INSTANCE_CREATE).unwrap(),
+            id_routes: vec!["/by-id/instances/{id}"],
+        },
+        // Lookup the previously created NIC
+        SetupReq::Get {
+            url: &*DEMO_INSTANCE_NIC_URL,
+            id_routes: vec!["/by-id/network-interfaces/{id}"],
         },
```
This is the (first) usecase of somewhere we may lookup a resource instead of creating a new resource mainly for the purpose of populating the `id` param.

zephraph commented on nexus/tests/integration_tests/unauthorized.rs:
```diff
@@ -288,6 +337,26 @@ async fn verify_endpoint(
+                    .to_string()
+                    .as_str(),
+            ),
+            None => endpoint
+                .url
+                .replace("{id}", "00000000-0000-0000-0000-000000000000"),
```
In this case there was no entry in `setup_requests` for some given endpoint that included `{id}`. This is true in the case of unimplemented endpoints like `images` and `snapshots` right now.

karencfv commented on nexus/tests/output/nexus_tags.txt:
```diff
@@ -4,6 +4,7 @@ disk_create                              /organizations/{organization_name}/proj
 disk_delete                              /organizations/{organization_name}/projects/{project_name}/disks/{disk_name}
 disk_list                                /organizations/{organization_name}/projects/{project_name}/disks
 disk_view                                /organizations/{organization_name}/projects/{project_name}/disks/{disk_name}
+disk_view_by_id                          /by-id/disks/{id}
```
What do you think about flipping the paths around like this:

```suggestion
disk_view_by_id                         /disks/by_id/{id}
```

It follows path convention a little closer and it would make command generation a little easier.

davepacheco replied:
Would the idea be that you'd have `/disks`, `/projects`, `/instances`, and all the others at the top level, and each of these would contain only one HTTP resource beneath them which was `by_id`?

karencfv replied:
I'm not sure if we are planning to have more global actions on resources? For example, we might want to have batch actions in the future where several resources are acted on.

In that case having the `by_id` bit of the path wouldn't make too much sense any more, the way I'm proposing 🤔 

Also, we'll probably want to have an endpoint for attaching disks with an ID as well. That would make it `instances/{id}/disks/attach/{id}`

I'm OK with either way we go :) If the majority prefer having `by_id` at the top level I'm ok with that :)

zephraph replied:
As for batching, I think the current endpoint could still support that. Something like `/by-id/{resource}?ids[]=1,2,3,4` is a rough approximation of how it could work.

karencfv replied:
Ha! I had not seen https://github.com/oxidecomputer/omicron/issues/1383. Sounds good to me then :)

davepacheco commented on nexus/src/external_api/http_entrypoints.rs:
```diff
@@ -3756,15 +4044,3 @@ async fn session_sshkey_delete(
     };
     apictx.external_latencies.instrument_dropshot_handler(&rqctx, handler).await
 }
-
```
What's up with the removal of this test?

zephraph replied:
I'm not sure how that happened exactly. Good catch.

zephraph replied:
Fixed now.

davepacheco commented on nexus/tests/integration_tests/unauthorized.rs:
```diff
@@ -114,13 +137,20 @@ G GET  PUT  POST DEL  TRCE G  URL
-    /// body of the `POST` request
-    body: serde_json::Value,
+/// The setup phase takes a list of `SetupReq` enums and issues a `GET` or `POST`
+/// request to each one's `url`. The `id_routes` field is a list of URLs that result
```
typo here? ("that result the result")

zephraph replied:
Cool, yeah. I improved it.

davepacheco commented on nexus/tests/integration_tests/unauthorized.rs:
```diff
@@ -121,6 +129,8 @@ struct SetupReq {
     /// body of the `POST` request
     body: serde_json::Value,
+    /// populates `id` params for later route lookups
+    id_routes: Vec<&'static str>,
```
This seems like a great simplification.

davepacheco reviewed (commented):
Nice!  This looks like the sort of thing that looks simple and clean now but probably involved a lot of grind to get here.  Thanks for doing it!

In regards to your specific feedback:

> Should the resource names in the route stay plural or be changed to be singular?

I'd go with plural, and that matches the existing stuff.

> I've spread the id implementations to be co-located near the associated GET method. Does that feel right or should I move back to having all the /by-id/ routes grouped together?

What you've got feels right to me.

ahl reviewed (commented):
I realized that I neglected to click submit earlier. Belated, but LGTM!! Nice work
